feat ($http): Add custom createXhr factory to config object

sjurba opened the pull request:
To be able to construct non-standard or customised XHR objects a factory an optional factory function can be added to the config object. This is required for use of the mozAnon and mozSystem options for FirefoxOS.
See https://developer.mozilla.org/en-US/docs/DOM/XMLHttpRequest#XMLHttpRequest() for details.

This is an alternative solution to the one suggested in PR  #7903. Closes #2318

sjurba committed:
feat ($http): Add custom createXhr factory to config object

Closes #2318

coolaj86 commented:
Bump. #fxos still needs this. See http://stackoverflow.com/a/29960399/151312

It's an easy enough workaround, but there are a decent number of newbies for whom it adds barrier to entry for using angular.

DaAwesomeP commented:
+1 I install Angular via bower and modifying the file after bower installs it is inconvenient and bower_components shouldn't be saved to a git repository anyways.

frekele commented:
+++1 !!!!!

caitp commented:
yeah, sorry this seems to have fell off the radar, I'll try to bring this up in the meeting tomorrow and put this back on track for landing

pkozlowski-opensource commented:
OK, let's review this one! I will send an alternative proposal and then we will compare / decide.

sjurba commented:
I rebased this one onto master. I also removed the refactoring of the httpbackend arguments since from the feedback it didn't seem like something we could do. 

So this is straight up attempt at adding an extra parameter to the config object. 

If the new parameter should only be available configurable on the global it's a simple fix to move it out of the extend function and always overwrite it. However if this is the way to go, there is maybe a simpler solution. Do we have/can we get access to the defaults object in the httpBackend somehow? If so it might simplify the implementation a bit.

I'm more than happy to contribute, but I'll need your guidance.

pkozlowski-opensource commented:
@sjurba we've discussed this PR during out last team meeting and I would like to run an alternative idea through you / the community. 

So, I understand what you are trying to do and your approach is simple and effective. Still there is one thing I'm not too keen on - adding more properties to the $http's config object. This makes $http's API surface larger - we've got a config object with gazillion of properties. This starts to resemble the directive definition object and we all know that this is not the nicest API.

The alternative idea would be to abstract XHR object creation in a dedicated service (provider). Give me sec, I'm going to open a sketch of the PR with this idea.

gkalpak commented:
:+1: for not overloading the config object. `$http`'s config **IS** the new DDO.

(I mean, I don't have to look up DDO properties any more (since several months now), but I still have to look up `$http`'s config every time :smile:)

pkozlowski-opensource commented:
@sjurba @gkalpak OK, here is an alternative approach: https://github.com/pkozlowski-opensource/angular.js/commit/249efc515c0a52321e146f9bc7b287074915e62d

Didn't test, missing unit-tests and docs etc. but this is just to get the idea across. Please let me know what do you think.

//cc: @petebacondarwin

gkalpak commented:
@pkozlowski-opensource, I like the general idea (although your implementation has some bugs :stuck_out_tongue:).

pkozlowski-opensource commented:
@gkalpak this is "pseudo implementation" :-) Want to confirm the approach before investing more than 5 minutes into it :-)

sjurba commented:
@pkozlowski-opensource I like it too. Unless someone else jumps to it, I'll be happy to give it a go when I have some time later this week.

pkozlowski-opensource commented:
@sjurba OK, I would be more than happy to let you continue along those lines! Feel free to ping me if you need any info / help!

DaAwesomeP commented:
Thanks!
